glusterd2/commands/volumes/volume-profile.go

```diff
@@ -0,0 +1,309 @@
+					brickProfileInfo.IntervalStats = populateStatsWithFop(brickProfileInfo.IntervalStats, key, value, intervalType)
+				}
+			}
+
+			brickProfileInfo.CumulativeStats = calculatePercentageLatencyForEachFop(brickProfileInfo.CumulativeStats)
+
+			brickProfileInfo.IntervalStats = calculatePercentageLatencyForEachFop(brickProfileInfo.IntervalStats)
+
+			// Append each brick's profile Info in an array and return the array
+			volumeProfileInfo = append(volumeProfileInfo, brickProfileInfo)
+		}
+	}
+
+	restutils.SendHTTPResponse(ctx, w, http.StatusOK, &volumeProfileInfo)
+}
+
+// Calculate Percentage latency for each fop in cumulative/interval stats
+func calculatePercentageLatencyForEachFop(stats StatType) StatType {
+	// Calculate sum of (hits * avgLatency) of all fops in Cumulative Stats. Used Later to calculate
+	// %-Latency for each FOP.
+	var tmpPercentageAvgLatency float64
+	for key := range stats.StatsInfo {
+		tmpAvgLatency, _ := strconv.ParseFloat(stats.StatsInfo[key]["avglatency"], 64)
+		tmpHits, _ := strconv.ParseFloat(stats.StatsInfo[key]["hits"], 64)
+		tmpPercentageAvgLatency += tmpAvgLatency * tmpHits
+	}
+	// Calculate %-Latency for each fop
+	// %-Avg-Latency  for one fop = 100 * (Hits for that fop * AvgLatency for that fop ) / (sum of hits * avgLatency of all fop)
+	for key := range stats.StatsInfo {
+		tmpAvgLatency, _ := strconv.ParseFloat(stats.StatsInfo[key]["avglatency"], 64)
+		tmpHits, _ := strconv.ParseFloat(stats.StatsInfo[key]["hits"], 64)
+		tmpPercentageAvgLatencyForEachFop := 100 * ((tmpAvgLatency * tmpHits) / tmpPercentageAvgLatency)
+		stats.StatsInfo[key]["%-latency"] = fmt.Sprintf("%f", tmpPercentageAvgLatencyForEachFop)
+	}
+	stats.PercentageAvgLatency = tmpPercentageAvgLatency
+
+	return stats
+}
+
+// populateStatsWithFop populates both interval/cumulative stats with fops and its profile info
+func populateStatsWithFop(stats StatType, key string, value string, ktype keyType) StatType {
+	var fop string
+	var k string
+	// Decode the fop and stat for the given key
+	switch ktype {
+	case cumulativeType:
+		fop, k = decodeCumulativeKey(key)
+		break
+	case intervalType:
+		fop, k = decodeIntervalKey(key)
+		break
+	}
+	if fop != "" && fop != "NULL" {
+		if _, ok := stats.StatsInfo[fop]; ok {
+			stats.StatsInfo[fop][k] = value
+		} else {
+			// Create new map with fop as key if this particular Fop is encountered for the first time
+			stats.StatsInfo[fop] = make(map[string]string)
+			stats.StatsInfo[fop][k] = value
+		}
+	} else {
+		if k == "read" {
+			stats.DataRead = value
+		} else if k == "write" {
+			stats.DataWrite = value
+		} else if k == "duration" {
+			stats.Duration = value
+		}
+	}
+	return stats
+}
+
+// Decode key for fop and stat. keys starting with -1 belong to cumulative stat
+// Eg: -1-12-maxlatency fop : 12        stat: maxlatency
+// -1-duration          fop : ""        stat: duration
+func decodeCumulativeKey(key string) (string, string) {
```

Comment:
It seems to me these functions are parsing the strings in much hard coded manner with lot of assumptions to the string lengths and all, which is bound to break later.